### ISSUE TITLE
Regression for `@classmethod` type variable inference in 1.1.354

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#7489]
```python
from __future__ import annotations

from typing import Generic, TypeVar, TypeVarTuple

Shape = TypeVarTuple("Shape")
DType = TypeVar("DType")
Dim = TypeVar("Dim", bound=int)


class ndarray(Generic[*Shape, DType]):  # noqa: N801
    ...


class Matrix1(Generic[Dim, *Shape, DType]):
    @classmethod
    def from_np(cls, a: ndarray[Dim, *Shape, DType]) -> Matrix1[Dim, *Shape, DType]:
        raise NotImplementedError


class Matrix2(Generic[*Shape, DType]):
    @classmethod
    def from_np(cls, a: ndarray[*Shape, DType]) -> Matrix2[*Shape, DType]:
        raise NotImplementedError


class Matrix3(Generic[Dim, DType]):
    @classmethod
    def from_np(cls, a: ndarray[Dim, DType]) -> Matrix3[Dim, DType]:
        raise NotImplementedError


def from_np(x: ndarray[Dim, *Shape, DType]) -> Matrix1[Dim, *Shape, DType]:
    raise NotImplementedError


def fn(x: ndarray[Dim, DType]):
    a = Matrix1.from_np(x) # Type error here
    reveal_type(a)
    b = Matrix2.from_np(x)
    reveal_type(b)
    c = Matrix3.from_np(x)
    reveal_type(c)
    d = from_np(x)
    reveal_type(d)
```

For the above code with pyright 1.1.353:

```
ml_scratch/type.py:38:17 - information: Type of "a" is "Matrix1[Dim@fn, DType@fn]"
ml_scratch/type.py:40:17 - information: Type of "b" is "Matrix2[Dim@fn, DType@fn]"
ml_scratch/type.py:42:17 - information: Type of "c" is "Matrix3[Dim@fn, DType@fn]"
ml_scratch/type.py:44:17 - information: Type of "d" is "Matrix1[Dim@fn, DType@fn]"
0 errors, 0 warnings, 4 informations
```

With pyright 1.1.354:

```
ml_scratch/type.py:37:25 - error: Argument of type "ndarray[Dim@fn, DType@fn]" cannot be assigned to parameter "a" of type "ndarray[Unknown, Unknown, Unknown]" in function "from_np"
    "ndarray[Dim@fn, DType@fn]" is incompatible with "ndarray[Unknown, Unknown, Unknown]"
      Type parameter "Shape@ndarray" is invariant, but "*tuple[Dim@fn]" is not the same as "*tuple[Unknown, Unknown]" (reportArgumentType)
ml_scratch/type.py:38:17 - information: Type of "a" is "Matrix1[Unknown, Unknown, Unknown]"
ml_scratch/type.py:40:17 - information: Type of "b" is "Matrix2[Unknown, Unknown]"
ml_scratch/type.py:42:17 - information: Type of "c" is "Matrix3[Unknown, Unknown]"
ml_scratch/type.py:44:17 - information: Type of "d" is "Matrix1[Dim@fn, DType@fn]"
1 error, 0 warnings, 4 informations
```

Note that the inferred type variables are now `Unknown` and that the `a = Matrix1.from_np(x)` line no longer type checks.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.355, which I just published.
